File: 2022/python/day03.py
"""
Script to find values of common items.
"""
from string import ascii_lowercase, ascii_uppercase

from day import Day

lowercase_map = {
            letter: ascii_lowercase.index(letter) + 1
            for letter in ascii_lowercase
        }

uppercase_map = {
    letter: ascii_uppercase.index(letter) + 27
    for letter in ascii_uppercase
}


class Day03(Day):
    def __init__(self):
        super().__init__()
        self.data = self.read_data()

    def solve_one(self) -> int:
        # Find the total value of common items, i.e. letters which appear in both halves of the string
        total_value = 0
        for line in self.data:
            midpoint = len(line) // 2
            first_half = set(line[:midpoint])
            second_half = set(line[midpoint:])

            # Find common letter between the two sets
            letter = first_half.intersection(second_half).pop()
            if letter.islower():
                total_value += lowercase_map[letter]
            else:
                total_value += uppercase_map[letter]

        return total_value

    def solve_two(self) -> int:
        # Find the total value of common items, i.e. letters which appear in each group of 3 strings
        total_value = 0
        for i in range(0, len(self.data), 3):
            string_1 = set(self.data[i])
            string_2 = set(self.data[i + 1])
            string_3 = set(self.data[i + 2])

            # Find common letter between the three sets
            letter = string_1.intersection(string_2).intersection(string_3).pop()
            if letter.islower():
                total_value += lowercase_map[letter]
            else:
                total_value += uppercase_map[letter]

        return total_value


def main():
    day03 = Day03()

    part_one_value = day03.solve_one()
    print(f"part one solution: {part_one_value}")

    part_two_value = day03.solve_two()
    print(f"part two solution: {part_two_value}")


if __name__ == "__main__":
    main()

File: 2022/java/Day03.java
/**
 * Script to find values of common items.
 */

import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import utils.Day;


public class Day03 extends Day {
    private final Map<Character, Integer> lowercaseMap;
    private final Map<Character, Integer> uppercaseMap;

    private Day03() {
        this.lowercaseMap = this.getLowercaseMap();
        this.uppercaseMap = this.getUppercaseMap();
    }

    private Map<Character, Integer> getLowercaseMap() {
        Map<Character, Integer> lowercaseMap;

        lowercaseMap = new HashMap<>();
        for (char c = 'a'; c <= 'z'; c++) {
            lowercaseMap.put(c, (int) c - 96);
        }

        return lowercaseMap;
    }

    private Map<Character, Integer> getUppercaseMap() {
        Map<Character, Integer> uppercaseMap;

        uppercaseMap = new HashMap<>();
        for (char c = 'A'; c <= 'Z'; c++) {
            uppercaseMap.put(c, (int) c - 38);
        }

        return uppercaseMap;
    }

    private List<String> parseData() {
        String value;
        List<String> values;

        // Read each line of input data as String, and store inside List
        values = new ArrayList<>();
        while (this.scanner.hasNextLine()) {
            value = this.scanner.nextLine();
            values.add(value);
        }

        return values;
    }

    private int solveOne(List<String> values) {
        int midpoint;
        int totalValue;
        String firstHalf;
        String secondHalf;
        Set<Character> firstSet;
        Set<Character> secondSet;
        Character letter;

        // Find the total value of common items, i.e. letters which appear in both halves of the string
        totalValue = 0;
        for (String value : values) {
            midpoint = value.length() / 2;
            firstHalf = value.substring(0, midpoint);
            secondHalf = value.substring(midpoint);

            firstSet = firstHalf.chars().mapToObj(c -> (char) c).collect(Collectors.toSet());
            secondSet = secondHalf.chars().mapToObj(c -> (char) c).collect(Collectors.toSet());

            // Find common letter between the two sets
            firstSet.retainAll(secondSet);
            letter = firstSet.iterator().next();

            if (Character.isLowerCase(letter)) {
                totalValue += this.lowercaseMap.get(letter);
            } else {
                totalValue += this.uppercaseMap.get(letter);
            }
        }

        return totalValue;
    }

    private int solveTwo(List<String> values) {
        int totalValue;
        String string1;
        String string2;
        String string3;
        Set<Character> firstSet;
        Set<Character> secondSet;
        Set<Character> thirdSet;
        Character letter;

        // Find the total value of common items, i.e. letters which appear in each group of 3 strings
        totalValue = 0;
        for (int i = 0; i < values.size(); i += 3) {
            string1 = values.get(i);
            string2 = values.get(i + 1);
            string3 = values.get(i + 2);

            firstSet = string1.chars().mapToObj(c -> (char) c).collect(Collectors.toSet());
            secondSet = string2.chars().mapToObj(c -> (char) c).collect(Collectors.toSet());
            thirdSet = string3.chars().mapToObj(c -> (char) c).collect(Collectors.toSet());

            // Find common letter between the two sets
            firstSet.retainAll(secondSet);
            firstSet.retainAll(thirdSet);
            letter = firstSet.iterator().next();

            if (Character.isLowerCase(letter)) {
                totalValue += this.lowercaseMap.get(letter);
            } else {
                totalValue += this.uppercaseMap.get(letter);
            }
        }

        return totalValue;
    }

    public static void main(String[] args) {
        Day03 day03;
        List<String> strings;
        int partOneValue;
        int partTwoValue;

        day03 = new Day03();
        strings = day03.parseData();

        partOneValue = day03.solveOne(strings);
        System.out.println("part one solution: " + partOneValue);

        partTwoValue = day03.solveTwo(strings);
        System.out.println("part two solution: " + partTwoValue);
    }
}
